sbin/mknod/mknod.8
.\"	$OpenBSD: src/sbin/mknod/mknod.8,v 1.17 2009/03/21 18:07:41 sobrado Exp $
.\"	$NetBSD: mknod.8,v 1.9 1995/08/10 23:47:32 jtc Exp $
.\"
.\" Copyright (c) 1980, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)mknod.8	8.2 (Berkeley) 12/11/93
.\"
.Dd $Mdocdate: January 6 2009 $
.Dt MKNOD 8
.Os
.Sh NAME
.Nm mknod
.Nd make device special files
.Sh SYNOPSIS
.Nm mknod
.Op Fl m Ar mode
.Ar name
.Op Cm b \*(Ba c
.Ar major minor
.Nm mknod
.Op Fl m Ar mode
.Ar name
.Cm p
.Sh DESCRIPTION
The
.Nm
command creates device special files.
Normally the shell script
.Pa /dev/MAKEDEV
is used to create special files for commonly known devices; it executes
.Nm
with the appropriate arguments and can make all the files required for the
device.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl m Ar mode
Set the file permission bits of newly created device special files to
.Ar mode .
The mode argument can be in any of the formats specified to the
.Xr chmod 1
utility.
If a symbolic mode is specified, the operators
.Ql +
and
.Ql -
are interpreted relative to an initial mode of
.Dq a=rw .
.El
.Pp
To make nodes manually, the arguments are:
.Bl -tag -width majorx
.It Ar name
Device or FIFO name.
For example
.Dq sd
for a SCSI disk or a
.Dq pty
for pseudo-devices.
FIFOs may be named arbitrarily by the user.
.It Cm b \*(Ba c \*(Ba p
Type of device or FIFO.
If the device is a block type device such as a tape or disk drive which needs
both cooked and raw special files,
the type is
.Cm b .
All other devices are character type devices, such as terminal
and pseudo devices, and are type
.Cm c .
A FIFO (also known as a named pipe) is type
.Cm p .
.It Ar major
The major device number is an integer number which tells the kernel
which device driver entry point to use.
To learn what major device number to use for a particular device,
check the file
.Pa /dev/MAKEDEV
to see if the device is known.
.It Ar minor
The minor device number tells the kernel which subunit
the node corresponds to on the device; for example,
a subunit may be a filesystem partition
or a tty line.
.Pp
Major and minor device numbers can be given in any format acceptable to
.Xr strtoul 3 ,
so that a leading
.Dq 0x
indicates a hexadecimal number, and a leading
.Dq 0
will cause the number to be interpreted as octal.
.El
.Sh SEE ALSO
.Xr chmod 1 ,
.Xr ksh 1 ,
.Xr mkfifo 1 ,
.Xr mkfifo 2 ,
.Xr mknod 2 ,
.Xr MAKEDEV 8
.Sh STANDARDS
.Nm
also exists as a built-in to
.Xr ksh 1 .
.Sh HISTORY
A
.Nm
command appeared in
.At v6 .


sbin/mknod/mknod.c
/*	$OpenBSD: src/sbin/mknod/mknod.c,v 1.16 2007/12/30 13:52:40 sobrado Exp $	*/
/*	$NetBSD: mknod.c,v 1.8 1995/08/11 00:08:18 jtc Exp $	*/

/*
 * Copyright (c) 1989, 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Kevin Fall.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static char copyright[] =
"@(#) Copyright (c) 1989, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)mknod.c	8.1 (Berkeley) 6/5/93";
#else
static char rcsid[] = "$OpenBSD: src/sbin/mknod/mknod.c,v 1.16 2007/12/30 13:52:40 sobrado Exp $";
#endif
#endif /* not lint */

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <locale.h>
#include <err.h>

extern char *__progname;

int domknod(char **, mode_t);
int domkfifo(char **, mode_t);
void usage(int);

int
main(int argc, char *argv[])
{
	int ch, ismkfifo = 0;
	void *set = NULL;
	mode_t mode = 0;

	setlocale (LC_ALL, "");

	if (strcmp(__progname, "mkfifo") == 0)
		ismkfifo = 1;

	while ((ch = getopt(argc, argv, "m:")) != -1)
		switch(ch) {
		case 'm':
			if (!(set = setmode(optarg))) {
				errx(1, "invalid file mode.");
				/* NOTREACHED */
			}

			/*
			 * In symbolic mode strings, the + and - operators are
			 * interpreted relative to an assumed initial mode of
			 * a=rw.
			 */
			mode = getmode(set, DEFFILEMODE);
			free(set);
			break;
		case '?':
		default:
			usage(ismkfifo);
		}
	argc -= optind;
	argv += optind;

	if (argv[0] == NULL)
		usage(ismkfifo);
	if (!ismkfifo) {
		if (argc == 2 && argv[1][0] == 'p') {
			ismkfifo = 2;
			argc--;
			argv[1] = NULL;
		} else if (argc != 4) {
			usage(ismkfifo);
			/* NOTREACHED */
		}
	}

	/*
	 * If the user specified a mode via `-m', don't allow the umask
	 * to modified it.  If no `-m' flag was specified, the default
	 * mode is the value of the bitwise inclusive or of S_IRUSR,
	 * S_IWUSR, S_IRGRP, S_IWGRP, S_IROTH, and S_IWOTH as modified by
	 * the umask.
	 */
	if (set)
		(void)umask(0);
	else
		mode = DEFFILEMODE;

	if (ismkfifo)
		exit(domkfifo(argv, mode));
	else
		exit(domknod(argv, mode));
}

int
domknod(char **argv, mode_t mode)
{
	dev_t dev;
	char *endp;
	u_int major, minor;

	if (argv[1][0] == 'c')
		mode |= S_IFCHR;
	else if (argv[1][0] == 'b')
		mode |= S_IFBLK;
	else {
		errx(1, "node must be type 'b' or 'c'.");
		/* NOTREACHED */
	}

	major = (long)strtoul(argv[2], &endp, 0);
	if (endp == argv[2] || *endp != '\0') {
		errx(1, "non-numeric major number.");
		/* NOTREACHED */
	}
	minor = (long)strtoul(argv[3], &endp, 0);
	if (endp == argv[3] || *endp != '\0') {
		errx(1, "non-numeric minor number.");
		/* NOTREACHED */
	}
	dev = makedev(major, minor);
	if (major(dev) != major || minor(dev) != minor) {
		errx(1, "major or minor number too large");
		/* NOTREACHED */
	}
	if (mknod(argv[0], mode, dev) < 0) {
		err(1, "%s", argv[0]);
		/* NOTREACHED */
	}
	return(0);
}

int
domkfifo(char **argv, mode_t mode)
{
	int rv;

	for (rv = 0; *argv; ++argv) {
		if (mkfifo(*argv, mode) < 0) {
			warn("%s", *argv);
			rv = 1;
		}
	}
	return(rv);
}

void
usage(int ismkfifo)
{

	if (ismkfifo == 1)
		(void)fprintf(stderr, "usage: %s [-m mode] fifo_name ...\n",
		    __progname);
	else {
		(void)fprintf(stderr,
		    "usage: %s [-m mode] name [b | c] major minor\n",
		    __progname);
		(void)fprintf(stderr, "       %s [-m mode] name p\n",
		    __progname);
	}
	exit(1);
}
